lib/Platform/Intl/impl_icu/Collator.cpp

```diff
@@ -0,0 +1,592 @@
+    case UCOL_LESS:
+      return -1;
+    case UCOL_EQUAL:
+      return 0;
+    case UCOL_GREATER:
+      return 1;
+  }
+  llvm_unreachable("Invalid result from ucol_strcoll");
+}
+
+// https://tc39.es/ecma402/#sec-intl.collator.prototype.resolvedoptions
+Options Collator::resolvedOptions() noexcept {
+  Options finalResolvedOptions;
+  finalResolvedOptions.emplace(
+      constants::opt_name::locale, Option(resolvedLocale_));
+  finalResolvedOptions.emplace(
+      constants::opt_name::usage, Option(resolvedUsageValue_));
+  finalResolvedOptions.emplace(
+      constants::opt_name::sensitivity, Option(resolvedSensitivityValue_));
+  finalResolvedOptions.emplace(
+      constants::opt_name::ignorePunctuation,
+      Option(resolvedIgnorePunctuationValue_));
+  finalResolvedOptions.emplace(
+      constants::opt_name::collation, Option(resolvedCollationValue_));
+  finalResolvedOptions.emplace(
+      constants::opt_name::numeric, Option(resolvedNumericValue_));
+  finalResolvedOptions.emplace(
+      constants::opt_name::caseFirst, Option(resolvedCaseFirstValue_));
+
+  return finalResolvedOptions;
+}
+
+vm::CallResult<std::vector<std::u16string>> Collator::supportedLocalesOf(
+    vm::Runtime &runtime,
+    const std::vector<std::u16string> &locales,
+    const Options &options) noexcept {
+  return LocaleResolver::supportedLocales(runtime, locales, options);
+}
+
+bool Collator::isExtensionTypeSupported(
+    std::u16string_view extensionKey,
+    std::u16string_view extensionType,
+    const LocaleBCP47Object &localeBCP47Object) {
+  if (extensionKey == constants::extension_key::kf) {
+    return std::find(
+               std::begin(constants::opt_value::case_first::validCaseFirsts),
+               std::end(constants::opt_value::case_first::validCaseFirsts),
+               extensionType) !=
+        std::end(constants::opt_value::case_first::validCaseFirsts);
+  }
+  if (extensionKey == constants::extension_key::kn) {
+    return std::find(
+               std::begin(constants::opt_value::numeric::validNumerics),
+               std::end(constants::opt_value::numeric::validNumerics),
+               extensionType) !=
+        std::end(constants::opt_value::numeric::validNumerics);
+  }
+  if (extensionKey == constants::extension_key::co) {
+    // The Intl.Collator spec disallows "standard" and "search" as an
+    // extension type per:
+    // https://tc39.github.io/ecma402/#sec-intl-collator-internal-slots
+    if (extensionType == collationTypeStandard ||
+        extensionType == collationTypeSearch) {
+      return false;
+    }
+    // ICU4C returns alias / legacy type ids for types that were available
+    // before Unicode extension type defined its ids to be
+    // 3 - 8 alphanumeric characters. Map these alias / legacy type ids to
+    // Unicode extension type ids.
+    // https://github.com/unicode-org/cldr/blob/main/common/bcp47/collation.xml
+    static std::unordered_map<std::u16string_view, std::u16string_view>
```

Comment:
This is pretty costly to construct and lookup given that there are only four entries. A constexpr array of pairs should be good enough for this.

Reply:
Done in 3rd commit.